util/git/client.go

```diff
@@ -60,15 +66,24 @@ func NewFactory() ClientFactory {
 	return &factory{}
 }
 
-func (f *factory) NewClient(rawRepoURL, path, username, password, sshPrivateKey string, insecureIgnoreHostKey bool) (Client, error) {
+func (f *factory) NewClient(rawRepoURL, path, username, password, sshPrivateKey string, insecure bool) (Client, error) {
 	var creds Creds
 	if sshPrivateKey != "" {
-		creds = SSHCreds{sshPrivateKey, insecureIgnoreHostKey}
+		creds = SSHCreds{sshPrivateKey, insecure}
 	} else if username != "" || password != "" {
 		creds = HTTPSCreds{username, password}
 	} else {
 		creds = NopCreds{}
 	}
+
+	// We need a custom HTTP client for go-git when we want to skip validation
+	// of the server's TLS certificate (--insecure-ignore-server-cert). Since
+	// this change is permanent to go-git Client during runtime, we need to
+	// explicitly replace it with default client for repositories without the
+	// insecure flag set.
+	if IsHTTPSURL(rawRepoURL) {
+		gitclient.InstallProtocol("https", githttp.NewClient(getRepoHTTPClient(rawRepoURL, insecure)))
```

Comment:
I think  this will override http client for all repos in all git clients. I had to come out with this ugly workaround to solve it: https://github.com/argoproj/argo-cd/pull/1890/files#diff-91bc9daf641d929e791c1b889c32a902

Reply:
Thanks for the hint, I will check for (and explicitly test for) this case. Since I think that every Git repo operation within ArgoCD will use NewClient() with the appropriate parameter set, it should be alright tho. 

getRepoHTTPClient() returns a configured instance of http.Client() according to its repository parametrization, so either with InsecureSkipVerify set to true in the TLS client config, with a CertPool containing CA certs in case the repo has dedicated verification configured or a plain, non-custom one (InsecureSkipVerify is false, system CA cert verification only). So, each Git repository connection should get its own, properly configured http.Client() object.

Reply:
Unfortunately `InstallProtocol` updates global variable which is used by all clients: [plumbing/transport/client/client.go](https://github.com/src-d/go-git/blob/master/plumbing/transport/client/client.go#L16). So it affects all clients, even ones which already exist. Also  `InstallProtocol` updates global map which might cause panic since concurrent map writes/reads are prohibited in golang.

Reply:
I think I understood the issue, and you are right. There is a chance of a race-condition happening, which leads in bad cases to non-accessible repositories (if InstallProtocol() was called a second time with different client settings before the previous client makes the actual connection to Git server), in worst case to panic if InstallProtocol() is called concurrently.

I guess I could integrate your workaround into my PR and replace your `insecureClient` construction with the value returned from the `getRepoHTTPClient()` function in my addition to `util/git/client.go`, because it also supports the addition of custom CA certificates for server verification. Will integrate your additional timeouts along the way.

Reply:
Or maybe the go-git people accept a PR for having a per-client object list of protocols, or at least a single protocol specified.

Reply:
Sure, please feel free to incorporate changes. Most of the code is in https://github.com/argoproj/argo-cd/pull/1890/files#diff-91bc9daf641d929e791c1b889c32a902

I hope so too. Created ticket to add the support: https://github.com/src-d/go-git/issues/1177 .

Reply:
I incorporated & tested your workaround, I just had to change the semantics a little - I don't pass the Insecure option through the Auth objects, because there might be repositories with a NopAuth as well. Instead, I extended the `nativeGitClient` struct to pass that option around (I found that to be much easier).